src/main/java/net/minecraftforge/client/MinecraftForgeClient.java

```diff
@@ -123,6 +129,20 @@ public static void onRebuildChunk(Level world, BlockPos position, RenderChunkReg
             regionCache.put(Pair.of(world, position), Optional.of(cache));
     }
 
+    private static final Map<ItemStack, LazyOptional<IItemDecoratorHandler>> customItemDecorationCache = new HashMap<>();
+    public static void onItemDecorations(Font font, ItemStack stack, int xOffset, int yOffset, @Nullable String stackSizeLabel)
+    {
+        LazyOptional<IItemDecoratorHandler> targetCapability = customItemDecorationCache.get(stack);
+        if (targetCapability == null)
+        {
+            ICapabilityProvider provider = stack;
+            targetCapability = provider.getCapability(CapabilityItemDecoratorHandler.ITEM_DECORATOR_HANDLER_CAPABILITY, null);
+            customItemDecorationCache.put(stack, targetCapability);
+            targetCapability.addListener(self -> customItemDecorationCache.put(stack, null));
+        }
+        targetCapability.ifPresent(decoration -> decoration.render(font, stack, xOffset, yOffset, stackSizeLabel));
+    }
```

Comment:
This cache is a huge memory leak.

Reply:
Agreed.  It should be some form of LRU cache, with a fixed upper bound on the size and auto-ejection of old entries.  Guava `LoadingCache`/`CacheBuilder` could work here.

Another problem is that ItemStack does not override `equals()` or `hashCode()`, so every separate instance of an ItemStack is a different map key, even if the two stacks are otherwise identical.  In other words:
```java
ItemStack stack = new ItemStack(Items.STONE);
cache.put(stack, handler);
LazyOptional<IItemDecoratorHandler> handler2 = cache.get(stack.copy());
```
doesn't work.  Unless that's your intention?